### DISCUSSION TITLE
How to CANCEL outgoing INVITE ?

### DISCUSSION THREAD

**Discussion by pbalash0v** [resiprocate/resiprocate#309] [Q&A] [answered]
Hello, the code below is roughly adapted from basicClient in dum tests.

```c++

//...create stack, create dum etc...
bool startedCallFlow = false;
bool startedCancel = false;
auto invite = std::shared_ptr<resip::SipMessage>{};
while (true)
{
	stack.process(50);
	while (dum->process())
	{
	}

	if (!startedCallFlow)
	{
		invite = dum->makeInviteSession(caller_aor, agent.m_sdp.get());
		dum->send(invite);
		startedCallFlow = true;
	}

	if (should_cancel) // somehow modified from outside at some point, initially false
	{
		if (!startedCancel)
		{
			if (auto adsh = dum->findAppDialogSet(resip::DialogSetId{*invite}); adsh.isValid())
			{
				adsh->end();
				startedCancel = true;				
			}
		}
	}
}
```
When the compiled program is run the callee is intentionally not reachable. Observing the sngrep output I can confirm that the `INVITE`s are keep getting retransmitted as expected. At some point `should_cancel` is set to `true` and than I expect `adsh->end()` to cause the INVITE transaction to terminate. But this doesn't happen, still observing INVITE retransmissions. Setting the log to `resip::Log::Stack` provides absolutely no hint. Any advice on what am I doing wrong here ?

Regards,
Pavel.

**Comment by sgodin** [comment, marked as answer]
An INVITE request cannot be CANCEL'd until a provisional response is received.  RFC3261 Section 9.1.

   _If no provisional response has been received, the CANCEL request MUST
   NOT be sent; rather, the client MUST wait for the arrival of a
   provisional response before sending the request.  
...
      If it was allowed to send the CANCEL before receiving a response
      for the previous request, the server could receive the CANCEL
      before the original request._

DUM flags that you "ended" the DialogSet, and will send a CANCEL if a provisional does arrive, otherwise you will need to wait until the INVITE transaction times out (32 seconds).

**Reply by pbalash0v** [reply]
Thanks a lot for a thorough explanation.
